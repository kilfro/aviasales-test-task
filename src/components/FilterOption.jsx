import '../style/filter-option.css'

import Checkbox from './Checkbox'
import PropTypes from 'prop-types'
import React from 'react'

const FilterOption = ({ value, ...rest }) => {
    return (
        <label htmlFor={rest.id} className='filter-option'>
            <Checkbox {...rest} />{value}
        </label>
    )
}

FilterOption.propTypes = {
    checked: PropTypes.bool.isRequired,
    id: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string.isRequired,
}

export default FilterOption