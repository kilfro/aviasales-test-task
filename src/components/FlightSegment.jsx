import { getDurationString, getFlightTimes, getStopsQuantityString } from '../functions'

import React from 'react'

const FlightSegment = ({ origin, destination, date, duration, stops }) => (
    <div className='flight-segment'>
        <div className='segment-part'>
            <div>{origin} - {destination}</div>
            <div>{getFlightTimes(new Date(date), duration)}</div>
        </div>
        <div className='segment-part'>
            <div>в пути</div>
            <div>{getDurationString(duration)}</div>
        </div>
        <div className='segment-part'>
            <div>{getStopsQuantityString(stops.length)}</div>
            <div>{stops.join(', ')}</div>
        </div>
    </div>
)

export default FlightSegment