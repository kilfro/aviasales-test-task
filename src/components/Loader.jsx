import '../style/loader.css'

import React from 'react'
import logo from '../logo.png'

const Loader = () => <img src={logo} className='loader' />

export default Loader