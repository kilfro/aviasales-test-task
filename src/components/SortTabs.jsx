import '../style/sort-tabs.css'

import React from 'react'
import { connect } from 'react-redux'
import { setOrderBy } from '../store/actions'

const SortTabs = ({ orderBy, setOrderBy }) => {
    const orderOptions = {
        price: 'Самый дешевый',
        duration: 'Самый быстрый',
    }

    const chosenStyle = {
        border: 'none',
        background: '#2196F3',
        color: '#FFFFFF',
    }

    const getChosenStyle = (option) => orderBy === option ? chosenStyle : {}

    return (
        <div className='tabs-container'>
            {Object.entries(orderOptions).map(([key, label]) => (
                <div
                    style={getChosenStyle(key)}
                    onClick={() => setOrderBy(key)}
                    key={key}
                >
                    {label}
                </div>
            ))}
        </div>
    )
}

const mapStateToProps = state => ({
    orderBy: state.orderBy
})

const mapDispatchToProps = {
    setOrderBy
}

export default connect(mapStateToProps, mapDispatchToProps)(SortTabs)