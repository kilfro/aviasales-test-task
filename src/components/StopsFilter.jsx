import '../style/stops-filter.css'

import FilterOption from './FilterOption'
import React from 'react'

const all = new Map()
all.set(0, 'Без пересадок')
all.set(1, '1 пересадка')
all.set(2, '2 пересадки')

const StopsFilter = ({ allOptions = all }) => {
    const [filterBy, setFilterBy] = React.useState([])
    const [chooseAll, setChooseAll] = React.useState(false)

    const isChecked = key => filterBy.includes(key)

    const changeHandler = key => {
        let newFilterBy
        if (isChecked(key)) {
            newFilterBy = filterBy.filter(f => f !== key)
            if (chooseAll) {
                setChooseAll(false)
            }
        } else {
            newFilterBy = [...filterBy, key]
            if (allOptions.size === newFilterBy.length) {
                setChooseAll(true)
            }
        }

        setFilterBy(newFilterBy)
    }

    const chooseAllHandler = () => {
        if (chooseAll) {
            setChooseAll(false)
            setFilterBy([])
        } else {
            setChooseAll(true)
            setFilterBy(Array.from(allOptions.keys()))
        }
    }

    return <div className='filter-container'>
        <div style={{margin: '0 20px 10px 20px'}}>количество пересадок</div>
        <FilterOption checked={chooseAll} onChange={chooseAllHandler} value='Все' />
        {Array.from(allOptions).map(([key, value]) => (
            <FilterOption id={key.toString()} checked={isChecked(key)} onChange={() => changeHandler(key)} value={value} key={key.toString()} />
        ))}
    </div>
}

export default StopsFilter