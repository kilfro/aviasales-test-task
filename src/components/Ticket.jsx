import '../style/ticket.css'

import FlightSegment from './FlightSegment'
import React from 'react'

const Ticket = ({ price, carrier, segments }) => {
    return (
        <div className='ticket-card'>
            <div className='ticket-header'>
                <div>{price.toLocaleString('ru-ru')} P</div>
                <img src={`http://pics.avs.io/99/36/${carrier}.png`} />
            </div>
            {segments.map(segment => <FlightSegment key={segment.date} {...segment} />)}
        </div>
    )
}

export default Ticket