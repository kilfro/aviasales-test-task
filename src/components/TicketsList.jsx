import '../style/tickets-list.css'

import React from 'react'
import Ticket from './Ticket'
import { connect } from 'react-redux'

const TicketsList = ({ tickets, stopsFilter }) => {
    const filterBySteps = ticket => (
        ticket.segments.reduce((sum, segment) => segment.stops.length === stopsFilter && sum, true)
    )

    return (
        <div className='tickets-list'>
            {tickets
                .filter(filterBySteps)
                .slice(0, 20)
                .map(ticket => <Ticket {...ticket} key={JSON.stringify(ticket)} />)}
        </div>
    )
}

const mapStateToProps = state => ({
    tickets: state.tickets,
    stopsFilter: state.stopsFilter,
})

export default connect(mapStateToProps)(TicketsList)