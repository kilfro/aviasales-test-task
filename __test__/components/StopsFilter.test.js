import FilterOption from '../../src/components/FilterOption'
import React from 'react'
import { StopsFilter } from '../../src/components/StopsFilter'

describe(StopsFilter, () => {
  test('must be rendered correctly', () => {
    const component = shallow(
      <StopsFilter filterBy={[]} setFilterBy={() => {}} />
    )

    expect(component).toMatchSnapshot()
  })

  test('must select all options and can el selection', () => {
    const setFilterHandler = jest.fn()
    const component = shallow(
      <StopsFilter filterBy={[]} setFilterBy={setFilterHandler} />
    )

    component.find(FilterOption).at(0).simulate('change')
    expect(setFilterHandler).toHaveBeenCalledWith([0, 1, 2, 3])

    component.find(FilterOption).at(0).simulate('change')
    expect(setFilterHandler).toHaveBeenCalledWith([])
  })

  test('must select one option', () => {
    const setFilterHandler = jest.fn()
    const component = shallow(
      <StopsFilter filterBy={[1]} setFilterBy={setFilterHandler} />
    )
    component.find(FilterOption).at(3).simulate('change')
    expect(setFilterHandler).toHaveBeenCalledWith([1, 2])
  })

  test('must set chooseAll state to "true"', () => {
    const setState = jest.fn()
    const useStateSpy = jest.spyOn(React, 'useState')
    useStateSpy.mockImplementation((init) => [init, setState])
    const component = shallow(
      <StopsFilter filterBy={[0, 1, 2]} setFilterBy={() => {}} />
    )

    component.find(FilterOption).last().simulate('change')
    expect(setState).toHaveBeenCalledWith(true)
  })

  test('must set chooseAll state to "false"', () => {
    const setState = jest.fn()
    const useStateSpy = jest.spyOn(React, 'useState')
    useStateSpy.mockImplementation(() => [true, setState])
    const component = shallow(
      <StopsFilter filterBy={[0, 1, 2, 3]} setFilterBy={() => {}} />
    )

    component.find(FilterOption).last().simulate('change')
    expect(setState).toHaveBeenCalledWith(false)
  })
})
