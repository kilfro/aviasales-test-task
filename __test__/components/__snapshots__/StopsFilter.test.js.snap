// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopsFilter must be rendered correctly 1`] = `
<div
  className="filter-container"
>
  <div
    style={
      Object {
        "margin": "0 20px 10px 20px",
      }
    }
  >
    количество пересадок
  </div>
  <FilterOption
    checked={false}
    id="all"
    onChange={[Function]}
    value="Все"
  />
  <FilterOption
    checked={false}
    id="0"
    key="0"
    onChange={[Function]}
    value="Без пересадок"
  />
  <FilterOption
    checked={false}
    id="1"
    key="1"
    onChange={[Function]}
    value="1 пересадка"
  />
  <FilterOption
    checked={false}
    id="2"
    key="2"
    onChange={[Function]}
    value="2 пересадки"
  />
  <FilterOption
    checked={false}
    id="3"
    key="3"
    onChange={[Function]}
    value="3 пересадки"
  />
</div>
`;
